File: Lab2/module.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <ctime>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include "module.h"

using namespace std;

void createTaskList(const char* fileName) {
	int numOfTasks, h, m;
	string s = "";
	Task task;

	ofstream outF(fileName, ios::binary);
	if (!outF) {
		cerr << "Error: file \"" << fileName << "\" could not be opened" << endl;
		exit(1);
	}

	cout << "Enter number of tasks: "; cin >> numOfTasks;
	cin.ignore();

	for (int i = 1; i <= numOfTasks; i++) {
		// task name
		do {
			printf("(%d) Task name: ", i); getline(cin, s);
		} while (s.size() > 100);

		strcpy(task.name, s.c_str());

		//start time
		do {
			cout << "Time of start (seaparate hours and minutes with space): "; cin >> h >> m; cin.ignore();
		} while (h < 0 || h >= 24 || m < 0 || m >= 60);

		task.startTime = h * 60 + m;

		//duration 
		do {
			cout << "Duration (seaparate hours and minutes with space): "; cin >> h >> m; cin.ignore();
		} while (h < 0 || h >= 24 || m < 0 || m >= 60);

		task.duration = h * 60 + m;
		task.endTime = task.startTime + task.duration;


		outF.write((char*)&task, sizeof(Task));
	}
	outF.close();
}

void displayTask(Task task) {
	printf("%15s:   %02d:%02d--->%02d:%02d--->%02d:%02d\n", task.name, task.startTime / 60, task.startTime % 60,
		task.duration / 60, task.duration % 60,
		task.endTime / 60, task.endTime % 60);
}

void displayTaskList(vector<Task> taskList) {
	Task task;
	for (size_t i = 0; i < taskList.size(); i++)
		displayTask(taskList.at(i));
}

vector<Task> getTaskList(const char* fileName) {
	Task task;
	vector<Task> taskList;

	ifstream inF(fileName, ios::binary);
	if (!inF) {
		cerr << "Error: file \"" << fileName << "\" could not be opened" << endl;
		exit(1);
	}

	while (inF.read((char*)&task, sizeof(task)))
		taskList.push_back(task);
	inF.close();
	return taskList;
}

bool compareTasks(Task t1, Task t2) {
	return (t1.startTime < t2.startTime);
}

void sortTaskList(vector<Task>& taskList) {
	sort(taskList.begin(), taskList.end(), compareTasks);
}

void findTheNearestTask(vector<Task> taskList) {
	//get current time in minutes after mindnight
	time_t curr_time = time(NULL);
	tm* tm_local = localtime(&curr_time);
	int currentTime = tm_local->tm_hour * 60 + tm_local->tm_min;

	//sorted vector
	Task task;
	for (size_t i = 0; i < taskList.size(); i++) {
		task = taskList.at(i);
		if (task.startTime > currentTime) {
			printf("\nCurrent time: %02d:%02d", tm_local->tm_hour, tm_local->tm_min);
			printf("\nThe next task is:\n");
			displayTask(task);
			return;
		}
	}
	printf("\nYou've got no task left\n");
}

void createSpareTimeList(vector<Task> taskList, const char* fileName) {
	Task currentTask, nextTask, spareTime;
	int a, b; //start & end of a task

	ofstream outF(fileName, ios::binary);
	if (!outF) {
		cerr << "Error: file \"" << fileName << "\" could not be opened" << endl;
		exit(1);
	}

	for (size_t i = 0; i < taskList.size() - 1; i++) {
		currentTask = taskList.at(i);
		nextTask = taskList.at(i + 1);

		a = currentTask.startTime;
		b = currentTask.endTime;
		if (b >= 13 * 60) { //if ends after 13:00

			if (a > 13 * 60) { //	check if there is spare time after 13:00, if so, add additional spare time interval
				spareTime = { "spare", 13 * 60, a - 13 * 60, a };
				outF.write((char*)&spareTime, sizeof(Task));
			}

			a = currentTask.endTime;
			b = nextTask.startTime;
			if (b - a > 0) {
				spareTime = { "spare", a, b - a, b };

				outF.write((char*)&spareTime, sizeof(Task));
			}
		}
	}

	b = taskList.at(taskList.size() - 1).endTime; //get the last task and check if there is spare time before 24:00
	if (b < 24 * 60) {
		spareTime = { "spare", b, 24 * 60 - b, 24 * 60 };

		outF.write((char*)&spareTime, sizeof(Task));
	}
	outF.close();
}

File: lab_2/module.py
from datetime import datetime as dt
class Task(object):
    name: str
    start_time: int
    duration: int
    end_time: int

def write_to_file(f, *args) -> None:
    '''Write args to the file one by one

    Parameters
    ----------
    file : <_io.BufferedWriter>
        Python file object to write
    args : list, optional
        Values to write 
    '''

    for arg in args:
        if isinstance(arg, str):
            arg = arg + '\0'*(100-len(arg)) #100 bytes
            f.write(bytes(arg.encode()))
        elif isinstance(arg, int):
            f.write(arg.to_bytes(4, byteorder='big')) #4 bytes

def create_task_list(file_name) -> None:
    '''Write a number of tasks to file
    The user is asked information about: number of tasks, 
    name of each task, start time of each task, duration of each task.
    Name, start time, duration and end time of the tasks are written
    
    Parameters
    ----------
    file_name : str
        The path to the file
    '''
    while True:
        try:
            n = int(input("Enter number of tasks: "))
        except ValueError: 
            pass
        else:
            break

    with open(file_name, 'wb') as f:
        for i in range(1, n+1):

            # task name
            while True:
                s = input(f"({i}) Task name: ")
                if len(s) < 100:
                    break

            # start time
            while True:
                try:
                    h, m = input("Time of start (seaparate hours and minutes with space): ").split(' ')
                    h, m = int(h), int(m)
                    if h >= 0 and h < 24 and m >= 0 and m < 60:
                        break
                except ValueError:
                    pass
                    
            start_time = h * 60 + m

            # duration
            while True:
                try:
                    h, m = input("Duration (seaparate hours and minutes with space): ").split(' ')
                    h, m = int(h), int(m)
                    if h >= 0 and h < 24 and m >= 0 and m < 60:
                        break
                except ValueError:
                    pass

            duration = h * 60 + m
            end_time = start_time + duration

            write_to_file(f, s, start_time, duration, end_time)

def display_task(task) -> None:
    '''Print out task information'''

    print("{:>15s}:   {:02d}:{:02d}--->{:02d}:{:02d}--->{:02d}:{:02d}".format(
          task.name, 
          task.start_time // 60, 
          task.start_time % 60,
		  task.duration // 60, 
          task.duration % 60,
		  task.end_time // 60, 
          task.end_time % 60));

def display_task_list(task_list) -> None:
    '''Print out information of tasks from task list'''

    for task in task_list:
        display_task(task)

def get_task_list(file_name) -> list:
    '''Read tasks from a file and create a list of them

    Parameters
    ----------
    file_name : str
        The path to the file to read

    Returns
    ----------
    task_list : list
        list of tasks
    '''

    task_list = []
    with open(file_name, 'rb') as f:
        while True:
            task = Task()
            task.name = f.read(100).decode().rstrip('\0')
            if not task.name: # if not eof
                break
            task.start_time = int.from_bytes(f.read(4), byteorder='big')
            task.duration = int.from_bytes(f.read(4), byteorder='big')
            task.end_time = int.from_bytes(f.read(4), byteorder='big')
            task_list.append(task)
    return task_list


def sort_task_list(task_list) -> list:
    '''Get sorted task lsit by the start time'''
    return sorted(task_list, key = lambda task : task.start_time)

def find_the_nearest_task(task_list) -> None:
    '''Find the nearest task after the current time
    Print out information about that task
    '''

    local_time = dt.now()
    current_time = local_time.hour*60 + local_time.minute
    
    for task in task_list:
        if task.start_time > current_time:
            print("\nCurrent time: {:02d}:{:02d}".format(
                local_time.hour, local_time.minute))
            print("The next task is:")
            display_task(task)
            return

    print("\nYou've got no task left")

def create_spare_time_list(task_list, spare_time_path) -> None:
    '''Write to the file a number of tasks in which each task represents the spare time after 13:00

    Parameters
    ----------
    task_list : list
        list of current tasks
    spare_time_path : str
        path to the file to write
    '''
    with open(spare_time_path, 'wb') as f:
        for i in range(len(task_list)-1):
            current_task = task_list[i]
            next_task = task_list[i+1] 

            a = current_task.start_time
            b = current_task.end_time
            if b >= 13*60: #if ends after 13:00
                if a > 13*60: #check if there is spare time after 13:00, if so, add additional spare time interval
                    write_to_file(f, "spare", 13*60, a - 13*60, a)

                a = current_task.end_time
                b = next_task.start_time
                if b - a > 0 : #get the last task and check if there is spare time before 24:00
                    write_to_file(f, "spare", a, b - a, b)

        b = task_list[-1].end_time
        if b < 24*60:
            if b > 13*60:
                write_to_file(f, "spare", b, 24*60 - b, 24*60)
            else:
                write_to_file(f, "spare", 13*60, 11 * 60, 24 * 60)
